hm_1/tasks/task_02_01.py

```diff
@@ -0,0 +1,72 @@
+"""
+Given a file containing text. Complete using only default collections:
+    1) Find 10 longest words consisting from largest amount of unique symbols
+    2) Find rarest symbol for document
+    3) Count every punctuation char
+    4) Count every non ascii char
+    5) Find most common non ascii char for document
+"""
+import string
+from collections import Counter
+from typing import List
+
+
+def get_data(file_path: str) -> str:
+    data = ""
+    with open(file_path, encoding="unicode_escape") as f:
+        for line in f:
+            data += line
+    return data
+
+
+def get_longest_diverse_words(file_path: str) -> List[str]:
+    data = {}
+    result = []
+    text_without_punctuation = get_data(file_path).translate(
+        str.maketrans("", "", string.punctuation)
+    )
+    words = text_without_punctuation.split()
+    for i in words:
+        data[i] = len(set(i.lower()))
+    k = len(data) if len(data) <= 10 else 10
+    result = sorted(data.items(), key=lambda x: x[1], reverse=True)
+    list_10_longest_words = []
+    for i in sorted(result[0:k], key=lambda x: len(x[0]), reverse=True):
+        list_10_longest_words.append(i[0])
+    return list_10_longest_words
+
+
+def get_rarest_char(file_path: str) -> str:
+    data = {}
+    text = get_data(file_path)
+    symbols = set(text.lower())
+    for i in symbols:
+        data[i] = text.lower().count(i)
+    sorted_data_by_rarest = sorted(data.items(), key=lambda k: k[0])
+    return sorted(sorted_data_by_rarest, key=lambda k: k[1])[0][0]
```

Comment:
What is the complexity of this function? Can we do better?

Reply:
for i in symbols:
        data[i] = text.lower().count(i) - O(n^2), 

sorted_data_by_rarest = sorted(data.items(), key=lambda k: k[0])- O(nlogn) =>O(n^2+nlogn), 
but in sorted() I’m not sure, I try to unite both sorted by amount of unique characters and by len(), but I wasn’t succeeded,  list(dict.items).sort() - I try too, but I don’t understand how to do it right, maybe can you give me a hint how to do it better?

Reply:
Let's start from the beginning. The key to the problem usually hidden in the problem.  What do you want to archive by this line of code?

Reply:
for i in symbols:
        data[i] = text.lower().count(i) - there I count how many times unique simbol goes in text and write item into dictionary with (‘unique symbol’ : how many times in a text),  after I sorted dictionary by value the by keys) I do it for avoid situation when some symbols have the same count of how many times in text, but I understand it isn’t working and I made it up use min(d, key.get) instead of sorted but I got the same result, Can I make test this function checking to exists symbol in possible list of values?

Reply:
Please pay attention to words. It really matters.  `by this line of code` is about one particular line.

`min` is what I expect because it is an O(n) operation. 

Unstable sorting is another problem. 
To solve it you can use `key` function that returns a tuple (value, key). This will order first by value and second by key and always produce the same ordering.